### PR TITLE
collections: make UserDict, UserList, and UserString concrete

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Part of #1476.

I added all methods that these classes implement in CPython (https://github.com/python/cpython/blob/master/Lib/collections/__init__.py).

### CODE REVIEW

File: stdlib/3/collections/__init__.pyi [hunk 1]
```diff
@@ -59,9 +59,133 @@ else:
...
+    @classmethod
+    def fromkeys(cls: Type[_UserDictT], iterable: Iterable[_KT], value: Optional[_VT] = ...) -> _UserDictT: ...
+
+_UserListT = TypeVar('_UserListT', bound=UserList)
+
+class UserList(MutableSequence[_T]):
+    def __init__(self, initlist: Optional[Iterable[_T]] = ...) -> None: ...
+    def __lt__(self, other: object) -> bool: ...
+    def __le__(self, other: object) -> bool: ...
+    def __eq__(self, other: object) -> bool: ...
```
Comment (matthiaskramm) [inline stdlib/3/collections/__init__.pyi]: If you override `__eq__`, should you overwrite `__ne__`, as well?

File: stdlib/3/collections/__init__.pyi [hunk 2]
```diff
@@ -59,9 +59,133 @@ else:
...
+    def __gt__(self, other: object) -> bool: ...
+    def __ge__(self, other: object) -> bool: ...
```
Comment (matthiaskramm) [inline stdlib/3/collections/__init__.pyi]: Could you add a comment where the comparison operators are declared abstract, necessitating the override here? I had difficulty finding that.

File: stdlib/3/collections/__init__.pyi [hunk 3]
```diff
@@ -59,9 +59,133 @@ else:
...
+_UserStringT = TypeVar('_UserStringT', bound=UserString)
+
+class UserString(Sequence[str]):
+    def __init__(self, seq: object) -> None: ...
+    def __str__(self) -> str: ...
+    def __repr__(self) -> str: ...
+    def __int__(self) -> int: ...
+    def __float__(self) -> float: ...
+    def __complex__(self) -> complex: ...
+    def __hash__(self) -> int: ...
```
Comment (matthiaskramm) [inline stdlib/3/collections/__init__.pyi]: Is it necessary to override `__hash__`?

**Comment by gvanrossum** [comment]
IIRC __ne__ is automatically supplied. But in fact all comparisons are
non-abstract methods of object so you shouldn't need to add any of these.
Please only add methods needed to make the classes concrete.

**Comment by JelleZijlstra** [comment]
Actually, object in typeshed doesn't declare any comparison operators. Mypy gives an error on `UserList() < UserList()` (which works at runtime) if I don't declare the comparison operator in the stub.

UserDict, UserList, and UserString actually implement numerous methods that aren't necessary to make them concrete, so I added those too.

I'll remove the redundant methods like `__eq__`, `__ne__`, `__hash__`, `__str__`, and `__repr__` though.

File: stdlib/3/collections/__init__.pyi (hunk 2 above)
Comment (gvanrossum) [inline stdlib/3/collections/__init__.pyi]: Hm, it's standard Python 3 behavior that `<`, `<=`, `>` and `>=` are not defined by default -- so they're not technically abstract, they're absent from `object`. However `==` and `!=` are present and concrete on `object`. I don't feel the need to add a comment here (though this is notably different from Python 2).

**Pull request merged**